### ISSUE TITLE
Can Espanso easily trigger Python scripts?

### ISSUE THREAD

**Issue by Raytle** [federico-terzi/espanso#708]
I've been using [AutoKey](https://github.com/autokey/autokey) for about 6 or 7 years. It's pretty good but crashes once in a while and is not being actively developed.

As I've learned a little Python over the last year or so I've been writing Python scripts and then simply triggering them with an abbreviation in AutoKey.

For example, in AutoKey I might associate `.foo1` with the following AutoKey script...

```
import runpy

runpy.run_path(path_name='~/my_python_scripts/foo1.py')
```

In other words, I type `.foo1` to trigger `~/my_python_scripts/foo1.py`. Can I do something similar in Espanso?

**Comment by risjain** [comment]
Of course you can - refer to https://espanso.org/docs/matches/#script-extension for more details. Though I have had some difficulties with getting to run my scripts using powershell/cmd, they run without a hiccup if you use shell: wsl. 

The primary difficulty (and I haven't explored much on why) was with passing variables between clipboard and the shell script to be called. This work out of the box, if you use wsl. 

If you already are aware of this, use espanso log to what is going on. Hope this helps.

**Comment by federico-terzi** [comment]
Hey @risjain,

Thank you for the help!

> The primary difficulty (and I haven't explored much on why) was with passing variables between clipboard and the shell script to be called. This work out of the box, if you use wsl.

That's interesting, as espanso pass those value as env variable on Windows as well. Perhaps, you forgot to use the CMD/Powershell specific way to access variables?
For example, to read the `EXAMPLE` env var on CMD you should use `echo %EXAMPLE%` and on Powershell `echo $env:EXAMPLE`

Cheers :)

**Comment by federico-terzi** [comment]
Feel free to reopen the issue if you have any other issue :)

**Comment by Raytle** [comment]
> Of course you can - refer to https://espanso.org/docs/matches/#shell-extension for more details. 

Thank you!

I just read,

>>> There will be tasks for which espanso was not designed for. For those cases, espanso offers the Script Extension, that enables you to call an external script, written in any language, and use its output in a match.

That seems like it will do what I want.

> Though I have had some difficulties with getting to run my scripts using powershell/cmd, they run without a hiccup if you use shell: wsl.

I see. Thanks for letting me know. I appreciate that.


> The primary difficulty (and I haven't explored much on why) was with passing variables between clipboard and the shell script to be called. This work out of the box, if you use wsl.

Hmmmm. I see. That's good to know.

 
> https://espanso.org/docs/matches/#script-extension didn't work for me as well for windows. But, maybe you can try.

Thanks for the information. I'm running Linux. 


> If you already are aware of this, use espanso log to what is going on. 

Ok. That sounds like a good idea.

> Hope this helps.

Yes it does. Thanks.

**Comment by Raytle** [comment]
> Feel free to reopen the issue if you have any other issue :)

Thanks for the offer. I will reopen this issue if necessary.
